distributed/tests/test_worker_memory.py

```diff
@@ -636,6 +636,100 @@ def leak():
     assert "memory" in out.lower()
 
 
+@gen_cluster(
+    nthreads=[("", 1)],
+    client=True,
+    worker_kwargs={"memory_limit": "10 GiB"},
+    config={
+        "distributed.worker.memory.target": False,
+        "distributed.worker.memory.spill": 0.7,
+        "distributed.worker.memory.pause": 0.9,
+        "distributed.worker.memory.monitor-interval": "10ms",
+    },
+)
+async def test_pause_while_spilling(c, s, a):
```

Comment:
Would it be possible to test not just that the paused state gets set, but that the pausing actually happens fast enough to prevent more work from running? (Corollary to my other comment.)

This is testing a race condition (how long tasks take to run vs the memory monitor interval).

For example, if you add a `sleep(memory_monitor_interval / 5)` to `SlowSpill.__init__` and submit N*2 tasks, it would be good to verify that between N and N + 5 tasks have _executed_ (possibly 1 fewer would be in memory) once we've paused. And the other ~N are waiting to run, but blocked by the pause.

Reply:
This sounds very brittle, as it would be heavily dependent on the responsiveness of the disk on the github CI.

Reply:
How about instead of `sleep(0.002)` in `SlowSpill.__init__`, which makes it hard to know exactly how many tasks will run before the memory monitor kicks in, we have a `threading.Semaphore(50)` that `SlowSpill.__init__` acquires? So we're guaranteed that exactly 50 `SlowSpill`s can be created.

Then I'm not following why it depends on disk? Nothing should be written to disk until we enter `_maybe_spill`. At that point there would be 50 `SlowSpill`s in `self.data` to evict.
1. We evict the first one. This immediately means the patched `get_process_memory` will indicate we're over the paused threshold. The worker isn't paused yet, so `__reduce__` sleeps for 100ms. Let's also add a `semaphore.release()` in `__reduce__`.
2. This single eviction takes longer than the `monitor-interval`. So we check `_maybe_pause_or_unpause`, and pause the worker. But we don't give control back to the event loop yet, because it hasn't been 0.5 sec.
4. The worker is now paused, so subsequent evictions don't sleep. If disk is fast, maybe we can evict all of them. If it's slow, we may return control the event loop multiple times before eviction is complete.

But regardless, since we paused the worker almost immediately, and the patched `get_process_memory` will never let it be un-paused, returning control to the event loop should be innocuous. We should be able to assert that no _new_ tasks started running. `len(a.data)` should be 50 or 51 (there might be 1 extra that started before pause, blocked on the semaphore, then completed during eviction), there should be no tasks running on the threadpool, and all the other tasks should still be in `a.ready`.

This wouldn't have been the case before https://github.com/dask/distributed/pull/6189/files#r857947472, that's why I'd like to test it.

Reply:
Ah true, that's annoying. I got this to work
```diff
diff --git a/distributed/tests/test_worker_memory.py b/distributed/tests/test_worker_memory.py
index c4ce4fd6..83d751f3 100644
--- a/distributed/tests/test_worker_memory.py
+++ b/distributed/tests/test_worker_memory.py
@@ -3,6 +3,7 @@ from __future__ import annotations
 import asyncio
 import logging
 from collections import Counter, UserDict
+import threading
 from time import sleep
 
 import pytest
@@ -649,8 +650,11 @@ async def test_nanny_terminate(c, s, a):
     },
 )
 async def test_pause_while_spilling(c, s, a):
+    N_PAUSE = 3
+    N_TOTAL = 5
+
     def get_process_memory():
-        if len(a.data) < 3:
+        if len(a.data) < N_PAUSE:
             # Don't trigger spilling until after all tasks have completed
             return 0
         elif a.data.fast and not a.data.slow:
@@ -663,25 +667,44 @@ async def test_pause_while_spilling(c, s, a):
     a.monitor.get_process_memory = get_process_memory
 
     class SlowSpill:
-        def __init__(self, _, sem: distributed.Semaphore):
-            self.sem = sem
-            # Block if there are 50 tasks in a.data.fast
-            sem.acquire()
+        # Can't pickle a `Semaphore`, so instead of a default value, we create it in `__init__`.
+        # Don't worry about race conditions; the worker is single-threaded.
+        sem: threading.Semaphore
+
+        def __init__(self, _):
+            try:
+                sem = self.sem
+            except AttributeError:
+                type(self).sem = threading.Semaphore(N_PAUSE)
+
+            # Block if there are `N_PAUSE` tasks in a.data.fast
+            self.sem.acquire()
 
         def __reduce__(self):
             paused = distributed.get_worker().status == Status.paused
+            self.sem.release()
             if not paused:
                 sleep(0.1)  # This is 10x the memory monitor interval
-            self.sem.release()
             return bool, (paused,)
 
-    sem = await distributed.Semaphore(3)
-    futs = c.map(SlowSpill, range(5), sem=sem)
-    while len(a.data.slow) < 3:
+    futs = c.map(SlowSpill, range(N_TOTAL))
+    while len(a.data.slow) < N_PAUSE:
         await asyncio.sleep(0.01)
     assert a.status == Status.paused
-    assert any(sp is True for sp in a.data.slow.values())
-    assert sum(ts.state == "ready" for ts in a.tasks.values()) == 2
+
+    assert not a.data.fast
+    assert len(a.data.slow) == N_PAUSE
+    # Worker should have become paused after the first `SlowSpill` was evicted,
+    # because the pickling (to write to disk) slept for longer than the memory monitor interval.
+    assert sum(sp is True for sp in a.data.slow.values()) == N_PAUSE - 1
+    # ^ NOTE: our hacked `__reduce__` means `SlowSpill` pickles into `bool(paused)`
+
+    # Due to pausing, no more tasks should have started. However, 1 might have already been executing
+    # (and is now complete) when the memory monitor kicked in.
+    assert sum(ts.state == "ready" for ts in a.tasks.values()) in (
+        N_TOTAL - N_PAUSE,
+        N_TOTAL - N_PAUSE - 1,
+    )
```

But I played with it a bit, and I can't get the test to fail. Changing the order of the `_maybe_pause_or_unpause` vs `sleep(0)`, adding more tasks, adding more sleeps in `__reduce__`, etc. The only thing that could break it was reordering the statements and changing the `sleep(0)` to a longer value (like `await sleep(0.5)`). So it seems like `sleep(0)` doesn't give much chance for other things to use the event loop.

I'm feeling like we probably don't need to test this.